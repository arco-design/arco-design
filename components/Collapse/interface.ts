import { CSSProperties, ReactNode } from 'react';

/**
 * @title Collapse
 */
export interface CollapseProps {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 当前面板选中值
   * @en Key of the active panel
   */
  activeKey?: string | string[];
  /**
   * @zh 默认展开的面板
   * @en Key of the default active panel
   */
  defaultActiveKey?: string | string[];
  /**
   * @zh 是否是手风琴模式
   * @en Whether to render as Accordion
   */
  accordion?: boolean;
  /**
   * @zh 自定义展开图标
   * @en Custom collapse icon
   */
  expandIcon?: ReactNode;
  /**
   * @zh 展开图标的位置
   * @en Position of collapse icon
   * @defaultValue left
   */
  expandIconPosition?: 'left' | 'right';
  /**
   * @zh 无边框样式
   * @en Whether to render border
   * @defaultValue true
   */
  bordered?: boolean;
  /**
   * @zh 设置为 `true` 时，挂载时不会渲染被隐藏的面板。
   * @en If true, invisible panels will not be rendered on mount
   * @defaultValue true
   */
  lazyload?: boolean;
  /**
   * @zh 是否销毁被折叠的面板
   * @en If true, panels will be unmounted on collapsing
   */
  destroyOnHide?: boolean;
  /**
   * @zh 可触发折叠操作的区域
   * @en The area that can trigger the collapse operation
   * @version 2.41.0
   */
  triggerRegion?: 'header' | 'icon';
  /**
   * @zh 展开面板改变时触发
   * @en Callback when the active panel changes
   */
  onChange?: (key: string, keys: string[], e) => void;
}

/**
 * @title Collapse.Item
 */
export interface CollapseItemProps {
  style?: CSSProperties;
  className?: string | string[];
  /**
   * @zh 内容区域的附加样式。
   * @en Additional styles of the content area.
   * @version 2.15.0
   */
  contentStyle?: CSSProperties;
  /**
   * @zh 折叠面板头部内容，允许自定义
   * @en Header content
   */
  header?: ReactNode;
  /**
   * @zh 对应 activeKey，当前面板组件的的唯一标识
   * @en Unique identifier key of the current panel item
   */
  name: string;
  /**
   * @zh 是否禁用
   * @en If true, the panel is not collapsible
   */
  disabled?: boolean;
  /**
   * @zh 自定义展开图标
   * @en Custom expand icon
   */
  expandIcon?: ReactNode;
  /**
   * @zh 是否展示展开按钮
   * @en Whether to show expand icon
   * @defaultValue true
   */
  showExpandIcon?: boolean;
  /**
   * @zh 额外节点
   * @en The extra element in the corner
   */
  extra?: ReactNode;
  /**
   * @zh 面板被折叠时是否销毁节点，优先级高于 `Collapse` 的 `destroyOnHide`
   * @en If true, item will be unmounted on collapsing. (Higher priority than `Collapse.destroyOnHide`)
   */
  destroyOnHide?: boolean;
}
