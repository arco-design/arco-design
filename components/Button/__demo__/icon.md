---
order: 1
title: 
  zh-CN: 图标按钮
  en-US: Icon
---

## zh-CN

Button 可以嵌入图标，在只设置图标而没有 children 时，按钮的高宽相等。

## en-US

Icons can be used in buttons. When `icon` is set and there are no children, the height and width of the button are equal.

```js
import { Button, Space } from '@arco-design/web-react';
import { IconPlus, IconDelete } from '@arco-design/web-react/icon';

const App = () => {
  return (
    <Space size="large">
      <Button type="primary" icon={<IconPlus />} />
      <Button type="primary" icon={<IconDelete />}>
        Delete
      </Button>
    </Space>
  );
};

export default App;
```
