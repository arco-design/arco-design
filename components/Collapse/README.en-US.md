`````
Component / Data Display

# Collapse

The content area that can be collapsed/expanded.
`````

%%Content%%

## API

### Collapse

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|accordion|Whether to render as Accordion|boolean |`-`|-|
|bordered|Whether to render border|boolean |`true`|-|
|destroyOnHide|If true, panels will be unmounted on collapsing|boolean |`-`|-|
|lazyload|If true, invisible panels will not be rendered on mount|boolean |`true`|-|
|expandIconPosition|Position of collapse icon|'left' \| 'right' |`left`|-|
|triggerRegion|The area that can trigger the collapse operation|'header' \| 'icon' |`-`|2.41.0|
|expandIcon|Custom collapse icon|ReactNode |`-`|-|
|activeKey|Key of the active panel|string \| string[] |`-`|-|
|className|Additional css class|string \| string[] |`-`|-|
|defaultActiveKey|Key of the default active panel|string \| string[] |`-`|-|
|style|Additional style|CSSProperties |`-`|-|
|onChange|Callback when the active panel changes|(key: string, keys: string[], e) => void |`-`|-|

### Collapse.Item

|Property|Description|Type|DefaultValue|Version|
|---|---|---|---|---|
|destroyOnHide|If true, item will be unmounted on collapsing. (Higher priority than `Collapse.destroyOnHide`)|boolean |`-`|-|
|disabled|If true, the panel is not collapsible|boolean |`-`|-|
|showExpandIcon|Whether to show expand icon|boolean |`true`|-|
|name|Unique identifier key of the current panel item|string  **(Required)**|`-`|-|
|expandIcon|Custom expand icon|ReactNode |`-`|-|
|extra|The extra element in the corner|ReactNode |`-`|-|
|header|Header content|ReactNode |`-`|-|
|className|Additional css class|string \| string[] |`-`|-|
|contentStyle|Additional styles of the content area.|CSSProperties |`-`|2.15.0|
|style|Additional style|CSSProperties |`-`|-|
