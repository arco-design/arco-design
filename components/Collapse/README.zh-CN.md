`````
组件 / 数据展示

# 折叠面板 Collapse

可以折叠 / 展开的内容区域。
`````

%%Content%%

## API

### Collapse

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|accordion|是否是手风琴模式|boolean |`-`|-|
|bordered|无边框样式|boolean |`true`|-|
|destroyOnHide|是否销毁被折叠的面板|boolean |`-`|-|
|lazyload|设置为 `true` 时，挂载时不会渲染被隐藏的面板。|boolean |`true`|-|
|expandIconPosition|展开图标的位置|'left' \| 'right' |`left`|-|
|triggerRegion|可触发折叠操作的区域|'header' \| 'icon' |`-`|2.41.0|
|expandIcon|自定义展开图标|ReactNode |`-`|-|
|activeKey|当前面板选中值|string \| string[] |`-`|-|
|className|节点类名|string \| string[] |`-`|-|
|defaultActiveKey|默认展开的面板|string \| string[] |`-`|-|
|style|节点样式|CSSProperties |`-`|-|
|onChange|展开面板改变时触发|(key: string, keys: string[], e) => void |`-`|-|

### Collapse.Item

|参数名|描述|类型|默认值|版本|
|---|---|---|---|---|
|destroyOnHide|面板被折叠时是否销毁节点，优先级高于 `Collapse` 的 `destroyOnHide`|boolean |`-`|-|
|disabled|是否禁用|boolean |`-`|-|
|showExpandIcon|是否展示展开按钮|boolean |`true`|-|
|name|对应 activeKey，当前面板组件的的唯一标识|string  **(必填)**|`-`|-|
|expandIcon|自定义展开图标|ReactNode |`-`|-|
|extra|额外节点|ReactNode |`-`|-|
|header|折叠面板头部内容，允许自定义|ReactNode |`-`|-|
|className|节点类名|string \| string[] |`-`|-|
|contentStyle|内容区域的附加样式。|CSSProperties |`-`|2.15.0|
|style|节点样式|CSSProperties |`-`|-|
