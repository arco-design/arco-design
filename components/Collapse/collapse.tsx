import React, { createContext, useContext, PropsWithChildren } from 'react';
import { isFunction } from '../_util/is';
import cs from '../_util/classNames';
import CollapseItem from './item';
import omit from '../_util/omit';
import { ConfigContext } from '../ConfigProvider';
import IconCaretRight from '../../icon/react-icon/IconCaretRight';
import useMergeValue from '../_util/hooks/useMergeValue';
import IconCaretLeft from '../../icon/react-icon/IconCaretLeft';
import { CollapseProps } from './interface';
import useMergeProps from '../_util/hooks/useMergeProps';

const getActiveKeys = (keys: CollapseProps['activeKey'], accordion: boolean): string[] => {
  const res = [].concat(keys);
  if (accordion) {
    return res.slice(0, 1);
  }
  return res;
};

const defaultProps: CollapseProps = {
  bordered: true,
  lazyload: true,
  expandIconPosition: 'left',
};

export const CollapseContext = createContext<
  Pick<
    CollapseProps,
    'expandIcon' | 'triggerRegion' | 'expandIconPosition' | 'lazyload' | 'destroyOnHide'
  > & {
    activeKeys: string[];
    onToggle?: (_key: string, _e) => void;
  }
>({
  expandIconPosition: 'left',
  expandIcon: <IconCaretRight />,
  activeKeys: [],
  onToggle: () => {},
});

function Collapse(baseProps: PropsWithChildren<CollapseProps>, ref) {
  const { getPrefixCls, componentConfig, rtl } = useContext(ConfigContext);
  const props = useMergeProps<PropsWithChildren<CollapseProps>>(
    baseProps,
    defaultProps,
    componentConfig?.Collapse
  );

  const [activeKeys, setActiveKeys] = useMergeValue<string[]>([], {
    defaultValue:
      'defaultActiveKey' in props
        ? getActiveKeys(props.defaultActiveKey, props.accordion)
        : undefined,
    value: 'activeKey' in props ? getActiveKeys(props.activeKey, props.accordion) : undefined,
  });

  const {
    children,
    className,
    style,
    bordered,
    lazyload,
    expandIcon,
    expandIconPosition,
    destroyOnHide,
    accordion,
    triggerRegion,
    onChange,
    ...rest
  } = props;

  const prefixCls = getPrefixCls('collapse');

  const onItemClick = (key: string, e): void => {
    let newActiveKeys = [...(activeKeys || [])];
    const index = activeKeys?.indexOf(key);
    if (index > -1) {
      newActiveKeys.splice(index, 1);
    } else if (accordion) {
      newActiveKeys = [key];
    } else {
      newActiveKeys.push(key);
    }
    if (!('activeKey' in props)) {
      setActiveKeys(newActiveKeys);
    }
    isFunction(onChange) && onChange(key, newActiveKeys, e);
  };

  return (
    <CollapseContext.Provider
      value={{
        activeKeys,
        triggerRegion,
        lazyload,
        destroyOnHide,
        expandIconPosition,
        onToggle: onItemClick,
        expandIcon:
          'expandIcon' in props ? (
            expandIcon
          ) : expandIconPosition === 'right' ? (
            <IconCaretLeft />
          ) : (
            <IconCaretRight />
          ),
      }}
    >
      <div
        ref={ref}
        {...omit(rest, ['activeKey', 'defaultActiveKey'])}
        className={cs(
          prefixCls,
          `${prefixCls}-${bordered ? 'border' : 'borderless'}`,
          { [`${prefixCls}-rtl`]: rtl },
          className
        )}
        style={style}
      >
        {children}
      </div>
    </CollapseContext.Provider>
  );
}

const ForwardRefCollapse = React.forwardRef<unknown, PropsWithChildren<CollapseProps>>(Collapse);

const CollapseComponent = ForwardRefCollapse as typeof ForwardRefCollapse & {
  Item: typeof CollapseItem;
};

CollapseComponent.displayName = 'Collapse';

CollapseComponent.Item = CollapseItem;

export default CollapseComponent;

export { CollapseProps };
