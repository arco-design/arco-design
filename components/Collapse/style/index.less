@import './token.less';
@import '../../style/mixin.less';

@collapse-prefix-cls: ~'@{prefix}-collapse';

.icon-hover(
  ~'@{collapse-prefix-cls}-item',
  @collapse-expand-icon-size,
  @collapse-expand-icon-size-bg
);

.icon-hover-bg(~'@{collapse-prefix-cls}-item', @collapse-expand-icon-color-bg);

.@{collapse-prefix-cls} {
  overflow: hidden;
  border-radius: @collapse-border-radius;
  border: @collapse-border-width solid @collapse-color-border;
  line-height: @collapse-line-height;

  &-rtl {
    direction: rtl;
  }

  &-item {
    border-bottom: @collapse-item-border-width solid @collapse-item-color-border;
    box-sizing: border-box;

    &-active {
      > .@{collapse-prefix-cls}-item-header {
        background-color: @collapse-title-color-bg_active;
        border-color: @collapse-title-color-border;
        transition: border-color 0s ease 0s;

        .@{collapse-prefix-cls}-item-header-title {
          font-weight: @collapse-title-font-weight_active;
        }
      }
    }

    &-header {
      display: flex;
      align-items: center;
      justify-content: space-between;
      position: relative;
      box-sizing: border-box;
      padding-top: @collapse-title-padding-vertical;
      padding-bottom: $padding-top;
      background-color: @collapse-title-color-bg;
      border-bottom: @collapse-title-border-width solid @color-transparent;
      color: @collapse-title-color-text;
      cursor: pointer;
      font-size: @collapse-title-font-size;
      overflow: hidden;
      line-height: @collapse-title-line-height;
      // 保证在收起动画完成时才隐藏边框
      transition: border-color 0s ease 0.19s;

      &[data-active-region='header'] {
        cursor: unset;

        > .@{collapse-prefix-cls}-item-header-title,
        > .@{prefix}-icon-hover {
          cursor: pointer;
        }
      }

      &[data-active-region='icon'] {
        cursor: unset;

        > .@{prefix}-icon-hover {
          cursor: pointer;
        }
      }

      &:focus-visible {
        box-shadow: inset 0 0 0 2px @collapse-color-box-shadow;
      }

      &-left {
        padding-left: @collapse-title-padding-horizontal + @collapse-expand-icon-size +
          @collapse-expand-icon-spacing-text;
        padding-right: @collapse-title-padding-horizontal;
      }

      &-right {
        padding-left: @collapse-title-padding-horizontal;
        padding-right: @collapse-title-padding-horizontal + @collapse-expand-icon-size +
          @collapse-expand-icon-spacing-text;
      }

      &-disabled {
        cursor: not-allowed;
        color: @collapse-title-color-text_disabled;
        background-color: @collapse-title-color-bg_disabled;

        .@{collapse-prefix-cls}-item-header-icon {
          color: @collapse-title-color-text_disabled;
        }
      }

      &-title {
        display: inline;
      }

      &-extra {
        float: right;
      }
    }

    // expand icon style
    & &-icon-hover {
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      left: @collapse-title-padding-horizontal;
      text-align: center;

      &-right {
        right: @collapse-title-padding-horizontal;
        left: unset;

        > .@{collapse-prefix-cls}-item-header-icon-down {
          transform: rotate(-90deg);
        }
      }
    }

    &-header-icon {
      color: @collapse-color-expand-icon;
      font-size: @collapse-expand-icon-size;
      transition: transform @transition-duration-2 @transition-timing-function-standard;
      vertical-align: middle;
      position: relative;
      display: block;

      &-down {
        transform: rotate(90deg);
      }
    }

    &-content {
      display: none;
      overflow: hidden;
      position: relative;
      color: @collapse-content-color-text;
      background-color: @collapse-content-color-bg;
      font-size: @collapse-content-font-size;
      // 展开动画时间使用 0.2s
      transition: height @transition-duration-2 @transition-timing-function-standard;

      &-expanded {
        display: block;
        height: auto;
      }

      &-box {
        padding: @collapse-content-padding-vertical @collapse-title-padding-horizontal
          @collapse-content-padding-vertical
          (
            @collapse-title-padding-horizontal + @collapse-expand-icon-size-bg +
              @collapse-expand-icon-spacing-text
          );
      }
    }

    &&-disabled > &-content {
      color: @collapse-content-color-text_disabled;
    }

    &-no-icon > &-header {
      padding-left: @collapse-title-padding-horizontal;
      padding-right: @collapse-title-padding-horizontal;
    }

    &:last-of-type {
      border-bottom: none;
    }
  }

  &.@{collapse-prefix-cls}-borderless {
    border: none;
  }

  &::after {
    display: table;
    content: '';
    clear: both;
  }
}
