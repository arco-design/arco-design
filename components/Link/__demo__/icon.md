---
order: 2
title:
  zh-CN: 图标
  en-US: Icon
---

## zh-CN
通过 `Icon` 属性设置带图标的链接，设置为 `true`时候显示默认图标。

## en-US

Customize icon node. If true, the default icon will be displayed.

```js
import { Link, Space } from '@arco-design/web-react';
import { IconEdit } from '@arco-design/web-react/icon';

const App = () => {
  return (
    <Space size={0} direction="vertical">
      <Space size="large">
        <Link href="#" icon>
          Hyperlinks
        </Link>
        <Link href="#" icon disabled>
          Hyperlinks
        </Link>
      </Space>
      <Space size="large">
        <Link href="#" icon={<IconEdit />}>
          Hyperlinks
        </Link>
        <Link href="#" icon={<IconEdit />} disabled>
          Hyperlinks
        </Link>
      </Space>
    </Space>
  );
};

export default App;
```
