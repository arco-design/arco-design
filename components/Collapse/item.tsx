import React, { PropsWithChildren, useContext } from 'react';
import { Transition } from 'react-transition-group';
import cs from '../_util/classNames';
import { CollapseContext } from './collapse';
import { ConfigContext } from '../ConfigProvider';
import IconHover from '../_class/icon-hover';
import { CollapseItemProps } from './interface';

function Item(props: PropsWithChildren<CollapseItemProps>, ref) {
  const { getPrefixCls } = useContext(ConfigContext);
  const ctx = useContext(CollapseContext);
  const {
    children,
    name,
    header,
    className,
    style,
    contentStyle,
    extra,
    disabled,
    destroyOnHide,
    expandIcon,
    showExpandIcon = true,
    ...rest
  } = props;

  const prefixCls = getPrefixCls('collapse-item');

  const isExpanded = ctx.activeKeys?.indexOf(name) > -1;
  const icon = showExpandIcon ? ('expandIcon' in props ? expandIcon : ctx.expandIcon) : null;

  return (
    <div
      ref={ref}
      {...rest}
      className={cs(
        prefixCls,
        {
          [`${prefixCls}-active`]: isExpanded,
          [`${prefixCls}-no-icon`]: !icon,
          [`${prefixCls}-disabled`]: disabled,
        },
        className
      )}
      style={style}
    >
      <div
        role="button"
        aria-disabled={disabled}
        aria-expanded={isExpanded}
        tabIndex={0}
        className={cs(`${prefixCls}-header`, `${prefixCls}-header-${ctx.expandIconPosition}`, {
          [`${prefixCls}-header-disabled`]: disabled,
        })}
        onClick={(e) => {
          !disabled && ctx.onToggle(name, e);
        }}
      >
        {icon && (
          <IconHover
            prefix={prefixCls}
            disabled={disabled}
            className={cs({
              [`${prefixCls}-icon-hover-right`]: ctx.expandIconPosition === 'right',
              [`${prefixCls}-header-icon-right`]: ctx.expandIconPosition === 'right',
            })}
          >
            <span
              className={cs(`${prefixCls}-header-icon`, {
                [`${prefixCls}-header-icon-down`]: isExpanded,
              })}
            >
              {icon}
            </span>
          </IconHover>
        )}
        <div className={`${prefixCls}-header-title`}>{header}</div>
        {extra && (
          <div
            className={`${prefixCls}-header-extra`}
            onClick={(e) => {
              e.stopPropagation();
            }}
          >
            {extra}
          </div>
        )}
      </div>
      <Transition
        in={isExpanded}
        addEndListener={(node, done) => {
          node.addEventListener('transitionend', done, false);
        }}
        mountOnEnter={'destroyOnHide' in props ? destroyOnHide : ctx.destroyOnHide || ctx.lazyload}
        unmountOnExit={'destroyOnHide' in props ? destroyOnHide : ctx.destroyOnHide}
        onEnter={(e) => {
          e.style.height = 0;
          e.style.display = 'block';
        }}
        onEntering={(e) => {
          e.style.height = `${e.scrollHeight}px`;
        }}
        onEntered={(e) => {
          e.style.height = 'auto';
        }}
        onExit={(e) => {
          e.style.display = 'block';
          e.style.height = `${e.offsetHeight}px`;
          // have to trigger reflow to get animation effect on exit
          e.offsetHeight; // eslint-disable-line
        }}
        onExiting={(e) => {
          e.style.height = 0;
        }}
        onExited={(e) => {
          e.style.display = 'none';
          e.style.height = 'auto';
        }}
      >
        <div
          role="region"
          className={cs(`${prefixCls}-content`, {
            [`${prefixCls}-content-expanded`]: isExpanded,
          })}
        >
          <div style={contentStyle} className={`${prefixCls}-content-box`}>
            {children}
          </div>
        </div>
      </Transition>
    </div>
  );
}

const ItemRef = React.forwardRef<unknown, PropsWithChildren<CollapseItemProps>>(Item);

ItemRef.displayName = 'CollapseItem';

export default ItemRef;

export { CollapseItemProps };
