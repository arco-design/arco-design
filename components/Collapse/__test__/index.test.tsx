import React from 'react';
import { mount } from 'enzyme';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Collapse from '../collapse';
import { IconInfoCircle, IconPlus } from '../../../icon';

mountTest(Collapse);
componentConfigTest(Collapse, 'Collapse');

const CollapseItem = Collapse.Item;

const prefixCls = '.arco-collapse';

const data = [
  {
    header: '琼恩·雪诺',
    content:
      '艾德公爵的私生儿子，与罗柏·史塔克同龄，兄弟关系亲密无间。其母身份不明，艾德公爵拒绝向任何人透露——有人传说那是南方某家族的一位贵族小姐，也有人说是一个寻常的使女。',
  },
  {
    header: '丹尼莉丝·坦格利安',
    content:
      '坦格利安王朝的末代君王伊里斯·坦格利安二世的幼女，银发紫眼，美貌异于常人。丹妮莉丝出生时一场剧烈的风暴袭击了龙石岛，“风暴降生”因此得名。',
  },
  {
    header: '提利昂·兰尼斯特',
    content:
      '泰温公爵和乔安娜夫人的第三个也是最小的孩子。因为是个侏儒，他有时候被戏称为小恶魔和半人。他利用自己的智慧屡次化险为夷，帮助兰尼斯特家族赢得了五王之战，但命运的不公使得他成为了一个弑亲者和通缉犯，踏上了流亡之路。',
  },
];

// eslint-disable-next-line
beforeAll(() => {
  //  https://github.com/yiminghe/css-animation/blob/a5986d73fd7dfce75665337f39b91483d63a4c8c/src/Event.js#L44
  // @ts-ignore
  window.AnimationEvent = window.AnimationEvent || (() => {});
  // @ts-ignore
  window.TransitionEvent = window.TransitionEvent || (() => {});
});

// eslint-disable-next-line
afterAll(() => {
  delete window.AnimationEvent;
  delete window.TransitionEvent;
});

describe('Collapse', () => {
  it('collapse render basic', () => {
    const wrapper = mount(
      <Collapse defaultActiveKey="1">
        {data.map((item, index) => (
          <CollapseItem key={index} header={item.header} name={index.toString()}>
            {item.content}
          </CollapseItem>
        ))}
      </Collapse>
    );

    expect(wrapper.find(`${prefixCls}-item`)).toHaveLength(data.length);
    expect(wrapper.find(`${prefixCls}-item`).at(1).hasClass(`arco-collapse-item-active`)).toBe(
      true
    );
  });

  it('collapse expandIcon', () => {
    const wrapper = mount(
      <Collapse>
        <CollapseItem key="1" header={data[0].header} expandIcon={<IconPlus />} name="2">
          {data[0].content}
        </CollapseItem>
        <CollapseItem header={data[1].header} showExpandIcon={false} name="1">
          {data[1].content}
        </CollapseItem>
      </Collapse>
    );

    expect(wrapper.find(`${prefixCls}-item`)).toHaveLength(2);
    expect(wrapper.find(`${prefixCls}-item`).at(0).find('IconPlus')).toHaveLength(1);
    expect(wrapper.find(`${prefixCls}-item`).at(1).find('IconPlus')).toHaveLength(0);
    expect(wrapper.find(`${prefixCls}-item`).at(1).hasClass('arco-collapse-item-no-icon')).toBe(
      true
    );
  });

  it('could be expand and collapse', () => {
    jest.useFakeTimers();
    const changeCollapse = jest.fn();
    const activeKeys = ['2'];
    const wrapper = mount(
      <Collapse defaultActiveKey={activeKeys} onChange={changeCollapse}>
        {data.map((item, index) => (
          <CollapseItem key={index} header={item.header} name={index.toString()}>
            {item.content}
          </CollapseItem>
        ))}
      </Collapse>
    );

    expect(wrapper.find(`${prefixCls}-item-active`)).toHaveLength(activeKeys.length);
    wrapper
      .find(`${prefixCls}-item-active`)
      .first()
      .find(`${prefixCls}-item-header`)
      .simulate('click');
    expect(wrapper.find(`${prefixCls}-item-active`)).toHaveLength(0);
    expect(changeCollapse.mock.calls).toHaveLength(1);

    for (let i = 0; i < data.length; i++) {
      wrapper.find(`${prefixCls}-item-header`).at(i).simulate('click');
    }

    expect(wrapper.find(`${prefixCls}-item-active`)).toHaveLength(data.length);
  });

  it('render correctly when content empty', () => {
    const wrapper = mount(
      <Collapse>
        {[].map((item, index) => (
          <CollapseItem key={index} header={item.header} name={index.toString()}>
            {item.content}
          </CollapseItem>
        ))}
      </Collapse>
    );
    expect(wrapper.find('Collapse')).toHaveLength(1);
    expect(wrapper.find('CollapseItem')).toHaveLength(0);
  });

  it('render self header correctly', () => {
    const wrapper = mount(
      <Collapse>
        {data.map((item, index) => (
          <CollapseItem
            key={index}
            header={
              <div>
                {item.header} <IconInfoCircle style={{ color: '#939aa3', marginLeft: 4 }} />{' '}
              </div>
            }
            name={index.toString()}
          >
            {item.content}
          </CollapseItem>
        ))}
      </Collapse>
    );
    expect(wrapper.find('IconInfoCircle')).toHaveLength(data.length);
  });

  it('expand and collapse correctly when accordion', () => {
    jest.useFakeTimers();
    let activeKey;
    const wrapper = mount(
      <Collapse
        accordion
        onChange={(_, keys) => {
          activeKey = keys;
        }}
      >
        {data.map((item, index) => (
          <CollapseItem key={index} header={item.header} name={String(index)}>
            {item.content}
          </CollapseItem>
        ))}
      </Collapse>
    );

    for (let i = 0; i < data.length; i++) {
      wrapper.find(`${prefixCls}-item-header`).at(i).simulate('click');
    }

    expect(activeKey).toHaveLength(1);
    expect(activeKey[0]).toEqual(String(data.length - 1));
  });

  it('expand and collapse correctly when set disabled', () => {
    jest.useFakeTimers();
    let activeKey;
    const changeCollapse = (_, keys) => {
      activeKey = keys;
    };
    const disabledIndex = 2;
    const wrapper = mount(
      <Collapse onChange={changeCollapse}>
        {data.map((item, index) => (
          <CollapseItem
            key={index}
            header={item.header}
            name={index.toString()}
            disabled={index === disabledIndex}
          >
            {item.content}
          </CollapseItem>
        ))}
      </Collapse>
    );
    for (let i = 0; i < data.length; i++) {
      wrapper.find(`${prefixCls}-item-header`).at(i).simulate('click');
    }

    expect(activeKey).toHaveLength(data.length - 1);

    expect(
      wrapper
        .find(`${prefixCls}-item-header`)
        .at(disabledIndex)
        .hasClass(`arco-collapse-item-header-disabled`)
    ).toBe(true);
  });

  it('render extra correctly', () => {
    const wrapper = mount(
      <Collapse expandIconPosition="right" expandIcon={<IconPlus />}>
        {data.map((item, index) => (
          <CollapseItem
            key={index}
            header={item.header}
            name={index.toString()}
            extra={<IconInfoCircle />}
          >
            {item.content}
          </CollapseItem>
        ))}
      </Collapse>
    );
    expect(wrapper.find('IconInfoCircle')).toHaveLength(data.length);
    expect(wrapper.find('IconPlus')).toHaveLength(data.length);
    expect(
      wrapper.find('.arco-collapse-item-header').at(0).hasClass('arco-collapse-item-header-right')
    );
  });

  it('render lazyload correctly', () => {
    const wrapper = mount(
      <Collapse lazyload defaultActiveKey={['0']}>
        {data.map((item, index) => (
          <CollapseItem
            key={index}
            header={item.header}
            name={index.toString()}
            extra={<IconInfoCircle />}
          >
            {item.content}
          </CollapseItem>
        ))}
      </Collapse>
    );
    expect(wrapper.find('.arco-collapse-item-content')).toHaveLength(1);
    wrapper.find(`${prefixCls}-item-header`).at(1).simulate('click');
    expect(wrapper.find('.arco-collapse-item-content')).toHaveLength(2);
  });
});
