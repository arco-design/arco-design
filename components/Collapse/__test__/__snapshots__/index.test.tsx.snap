// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Collapse default 1`] = `
<div
  class="arco-collapse arco-collapse-border"
/>
`;

exports[`ConfigProvider componentConfig.Collapse set className globally 1`] = `
<div
  class="arco-collapse arco-collapse-border global-default-class"
/>
`;

exports[`ConfigProvider componentConfig.Collapse set className globally and component set className itself. 1`] = `
<div
  class="arco-collapse arco-collapse-border component-class"
/>
`;

exports[`ConfigProvider componentConfig.Collapse set data-* property 1`] = `
<div
  class="arco-collapse arco-collapse-border"
  data-test-name="Collapse"
/>
`;
